Document HTMLAllCollection (#27974)
mdn/content: files/en-us/web/api/htmlallcollection/index.md

---
title: HTMLAllCollection
slug: Web/API/HTMLAllCollection
page-type: web-api-interface
browser-compat: api.HTMLAllCollection
---

{{APIRef("DOM")}}{{Deprecated_Header}}

> The **`HTMLAllCollection`** interface represents a collection of _all_ of the document's elements, accessible by order (like an array) and by ID (like a regular object). It is returned by the {{domxref("document.all")}} property.

Reviewer: Probably worth a short para about how this is almost identical to `HTMLCollection` (the spec is hilarious here):

> the main differences are that it allows a staggering variety of different (ab)uses of its methods to all end up returning something, and that it can be called as a function as an alternative to property access.

## Instance properties

- {{domxref("HTMLAllCollection.length")}} {{ReadOnlyInline}}
  - : Returns the number of items in the collection.

## Instance methods

- {{domxref("HTMLAllCollection.item()")}}

>   - : Returns the specific node at the given zero-based `index` into the list. Returns `null` if the `index` is out of range.

Author: Element. When you see "node" it's copied from `HTMLCollection`, which is wrong anyway.

- {{domxref("HTMLCollection.namedItem()")}}

>   - : Returns the specific node whose ID or, as a fallback, name matches the string specified by `name`. Matching by name is only done as a last resort, only in HTML, and only if the referenced element supports the `name` attribute. Returns `null` if no node exists by the given name.

Reviewer: ```suggestion
  - : Returns the first [element](https://developer.mozilla.org/en-US/docs/Web/API/Element) in the collection whose [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) or `name` attribute match the given string name, or `null` if no element matches.
```

(I copied this from the preamble to https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection/namedItem. Mainly, the existing para was too long and a bit vague (i.e. terms like "as a last resort", what does that mean?).

Author: Again, the current para is copied from the `HTMLCollection` landing page. It's all a bit messy, isn't it...

Reviewer: Yeah, the old DOM docs are pretty bad.

## Usage in JavaScript

### Indexed access

> In addition to the methods above, an `HTMLAllCollection` can be accessed by integer indices and string property names. HTML IDs may contain `:` and `.` as valid characters, which would necessitate using bracket notation for property access. `collection[i]` is equivalent to `collection.item(i)`, where `i` can be an integer, a string containing an integer, or a string representing an ID.

Reviewer: ```suggestion
In addition to the methods above, elements in an `HTMLAllCollection` can be accessed by integer indices and string property names. The HTML `id` attribute may contain `:` and `.` as valid characters, which would necessitate using bracket notation for property access. `collection[i]` is equivalent to `collection.item(i)`, where `i` can be an integer, a string containing an integer, or a string representing an `id`.
```

### Calling as a function

An `HTMLAllCollection` object is callable. When it's called with no arguments or with `undefined`, it returns `null`. Otherwise, it returns the same value as the {{domxref("HTMLAllCollection/item", "item()")}} method when given the same arguments.

### Special type conversion behavior

> For historical reasons, `document.all` is an object that in many ways behaves like `undefined`. Specifically:

Reviewer: ```suggestion
For historical reasons, `document.all` is an object that in the following ways behaves like `undefined`:
```
